src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,304 @@
+                        mark_ty(used_parameters, ty);
+                    }
+                }
+            }
+            ty::PredicateKind::Projection(predicate, ..) => {
+                let self_ty = predicate.skip_binder().projection_ty.self_ty();
+                if is_self_ty_used(used_parameters, self_ty) {
+                    let ty = predicate.ty();
+                    debug!("unused_generic_params: (projection) ty={:?}", ty);
+                    mark_ty(used_parameters, ty.skip_binder());
+                }
+            }
+            _ => (),
+        }
+    }
+}
+
+/// Emit an error for the function represented by `def_id`, labelling each generic parameter which
+/// was unused.
+fn emit_unused_generic_params_error<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    unused_parameters: u64,
+) {
+    debug!("emit_unused_generic_params_error: def_id={:?}", def_id);
+    if !def_id.is_local() {
+        return;
+    }
+
+    debug!("emit_unused_generic_params_error: unused_parameters={:064b}", unused_parameters);
+    let fn_span = match tcx.opt_item_name(def_id) {
+        Some(ident) => ident.span,
+        _ => tcx.def_span(def_id),
+    };
+
+    let mut err = tcx.sess.struct_span_err(fn_span, "item has unused generic parameters");
+
+    let mut next_generics = Some(generics);
+    while let Some(generics) = next_generics {
+        for param in &generics.params {
+            if is_bit_set(unused_parameters, param.index) {
+                debug!("emit_unused_generic_params_error: param={:?}", param);
+                let def_span = tcx.def_span(param.def_id);
+                err.span_label(def_span, &format!("generic parameter `{}` is unused", param.name));
+            }
+        }
+
+        next_generics = generics.parent.map(|did| tcx.generics_of(did));
+    }
+
+    err.emit();
+}
+
+/// Visitor used to aggregate generic parameter uses.
+struct UsedGenericParametersVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    used_parameters: &'a mut u64,
+}
+
+impl<'a, 'tcx> UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_foldable<T: std::fmt::Debug + TypeFoldable<'tcx>>(&mut self, value: T) {
+        debug!("visit_foldable: value={:?}", value);
+        // Exit early if this type has no type or const parameters.
+        if !value.has_param_types_or_consts() {
+            return;
+        }
+
+        // Use `TypeVisitor` to find any generic parameters.
+        value.visit_with(self);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_local_decl(&mut self, local: Local, local_decl: &LocalDecl<'tcx>) {
+        debug!("visit_local_decl: local_decl={:?}", local_decl);
+        let is_closure_or_generator =
+            self.tcx.is_closure(self.def_id) || self.tcx.type_of(self.def_id).is_generator();
+        if is_closure_or_generator && local == Local::from_usize(1) {
```

Comment:
Maybe match on `self.tcx.def_kind(self.def_id)`? But also, only do this if the local is local number 1, otherwise it's just repeating the query unnecessarily.

Reply:
Fixed!